tstime/tstime.go

```diff
@@ -6,117 +6,234 @@
+	return context.WithDeadline(ctx, t)
+}
+
+// Virtual is a Time implementation that runs faster or slower than
+// wall clock time, but keeps ratios the same.
+type Virtual struct {
+	// Second is how long a virtual second actually takes in wall
+	// clock time. That is, if you virtual.Sleep(time.Second), this is
+	// how long you'll actually sleep.
+	Second time.Duration
+
+	// first is the first value of time.Now seen by this virtual
+	// instance. It's used as the origin point for virtual time.Time
+	// values when scaling them to/from wall clock time.
+	first int64 // nanoseconds since Epoch, accessed atomically.
+}
+
+func (vt *Virtual) init() {
+	atomic.CompareAndSwapInt64(&vt.first, 0, time.Now().UnixNano())
+}
+
+// mulDiv implements v*mul/div using 128-bit integer math, to avoid
+// overflows and losses of precision.
+func mulDiv(v, mul, div time.Duration) time.Duration {
+	neg := v < 0
+	if neg {
+		v = -v
 	}
-	nsStr := s[baseLen:]
-	if nsStr != "" {
-		if nsStr[0] != '.' {
-			return time.Time{}, errors.New("invalid optional nanosecond prefix")
-		}
-		if !parseInt(nsStr[1:], &nsec) {
-			return time.Time{}, fmt.Errorf("invalid optional nanosecond number %q", nsStr[1:])
-		}
-		for i := 0; i < len("999999999")-(len(nsStr)-1); i++ {
-			nsec *= 10
-		}
+
+	hi, lo := bits.Mul64(uint64(v), uint64(mul))
+	quo, _ := bits.Div64(hi, lo, uint64(div))
+
+	ret := time.Duration(quo)
+	if neg {
+		ret = -ret
 	}
-	return time.Date(year, time.Month(mon), day, hr, min, sec, nsec, loc), nil
+	return ret
+}
+
+// virtualToWallDuration converts a time.Duration from the virtual
+// reference frame to the wall clock reference frame. This is the
+// transform that takes place to convert
+// virtual.Sleep(virtualDuration) into time.Sleep(realDuration), for
+// example.
+func (vt *Virtual) virtualToWallDuration(d time.Duration) time.Duration {
+	return mulDiv(d, vt.Second, time.Second)
+}
+
+// wallToVirtualDuration converts a time.Duration from the wall clock
+// reference frame to the virtual reference frame.
+func (vt *Virtual) wallToVirtualDuration(d time.Duration) time.Duration {
+	return mulDiv(d, time.Second, vt.Second)
+}
+
+func (vt *Virtual) virtualToWallTime(t time.Time) time.Time {
+	vt.init()
+	wt := time.Unix(0, atomic.LoadInt64(&vt.first)).In(t.Location())
+	return wt.Add(vt.virtualToWallDuration(t.Sub(wt)))
+}
+func (vt *Virtual) wallToVirtualTime(t time.Time) time.Time {
+	vt.init()
+	ret := time.Unix(0, atomic.LoadInt64(&vt.first)).In(t.Location())
+	return ret.Add(vt.wallToVirtualDuration(t.Sub(ret)))
+}
+
+// After is like time.After.
+func (vt *Virtual) After(d time.Duration) <-chan time.Time {
+	return time.After(vt.virtualToWallDuration(d))
+}
+
+// Sleep is like time.Sleep.
+func (vt *Virtual) Sleep(d time.Duration) {
+	time.Sleep(vt.virtualToWallDuration(d))
+}
+
+// Since is like time.Since.
+func (vt *Virtual) Since(t time.Time) time.Duration {
+	return vt.wallToVirtualDuration(time.Since(vt.virtualToWallTime(t)))
```

Comment:
`virtualToWallTime` breaks monotonic time support (see https://go.googlesource.com/proposal/+/master/design/12914-monotonic.md), so this version of time.Since will behave differently in the fact of machine time changes. Maybe that's ok if it's intended for tests, but it could cause mysterious test flakes.

I haven't worked it all the way through, but I believe that you could work around this with even more wrapping. You could rename your `Time` interface to just `T`, and then instead of having this package accept/return `time.Time`s, you could have them accept/return `type Time interfaces. Clock times are unchanged (modulo being wrapped), but Virtual methods return virtual times, which are a struct containing the virtual time and the clock. You'd need to modify your `T` interface methods to deal in `tstime.Time` instead of `time.Time`, and your Std implementation accordingly.

Then the implementation of this function looks like:

```go
if u, ok := t.(*virtualTime); !ok {
  panic("cannot mix and match different tstime.T")
}
return vt.wallToVirtualDuration(time.Since(u.Clock))
```

The panic in this function is another sign that the existing implementation is a little bit dangerous: Mixing time.Times between different tstime.Times will cause silent weird results, not failures.